src/rgw/rgw_datalog.h

```diff
@@ -125,6 +126,31 @@ struct RGWDataChangesLogMarker {
   RGWDataChangesLogMarker() = default;
 };
 
+struct rgw_data_notify_entry {
+  std::string key;
+  uint64_t gen = 0;
+
+  rgw_data_notify_entry(const std::string key, uint64_t gen): key(key), gen(gen) {}
+
+  rgw_data_notify_entry(const std::string& strKey = "")
+  : key(strKey) {}
```

Comment:
In RGWDataChangesLog::mark_modified(), I wanted to be able to check if the 'key' exists or not by using count(), so I have explicitly defined these.
`    if (shard != modified_shards.end() && shard->second.count(key)) {
      return;
    }`

Reply:
hmm, i see. `mark_modified()` uses that check to avoid taking an exclusive lock

but we should be looking for an entry matching `{key, gen}` there, not `{key, 0}`

presence of a `{"foo", 0}` entry shouldn't prevent us from inserting a `{"foo", 1}`